### PR TITLE
Add a doc on guidelines for contributing a new dialect to the MLIR core repo

### PR DESCRIPTION
**Pull request by joker-eph**

### CODE REVIEW

File: g3doc/DeveloperGuide.md [hunk 1]
```diff
@@ -51,3 +51,55 @@ pipeline.
...
+
+It is prefered to upstream your dialects/components in small incremental
+patches that can be individually reviewed. That is, after the initial RFC has
+been agreed on, we encourage dialects to be built progressively by faster
+iterations in-tree ; as long as it is clear they evolve towards their
+milestones and goals.
+
+We have seen the following broad categories of dialects:
+  * Edge dialects that model a representation external to MLIR. Examples include
+    LLVM, SPIR-V dialects, TensorFlow, XLA/HLO, … Such dialects may be a better
```
Comment (jpienaar) [inline g3doc/DeveloperGuide.md]: Perhaps avoid unicode?

File: g3doc/DeveloperGuide.md [hunk 2]
```diff
@@ -51,3 +51,55 @@ pipeline.
...
+ * Who are the future contributors/maintainers beyond those who propose the
+   dialect?
+
+On a practical aspect, we will expect the code to follow the other sections of
+this document, with an emphasis on the documentation alongside the source code.
+
+It is prefered to upstream your dialects/components in small incremental
+patches that can be individually reviewed. That is, after the initial RFC has
+been agreed on, we encourage dialects to be built progressively by faster
+iterations in-tree ; as long as it is clear they evolve towards their
```
Comment (jpienaar) [inline g3doc/DeveloperGuide.md]: Extra space (s/ ; /; / ?)

File: g3doc/DeveloperGuide.md [hunk 3]
```diff
@@ -51,3 +51,55 @@ pipeline.
...
+    other dialects or introduce a programming model. Generalization is sometimes
+    demonstrated by having several dialects lower to or originate from a new
+    dialect. While additional abstractions may be useful, they should be traded
+    off against the additional complexity of the dialect ecosystem. Examples of
+    abstraction dialects include the GPU and Loop dialects.
+  * Transformation dialects that serve as input/output for program
+    transformations. These dialects are commonly introduced to materialize
+    transformation pre- and post-conditions in the IR, while conditions can be
+    obtained through analysis or through operation semantics. Examples include
+    Affine, Linalg dialects.
```
Comment (jpienaar) [inline g3doc/DeveloperGuide.md]: s/,/ and/ ?

File: g3doc/DeveloperGuide.md (hunk 2 above)
Comment (joker-eph) [reply g3doc/DeveloperGuide.md]: French style :)

File: g3doc/DeveloperGuide.md (hunk 1 above)
Comment (joker-eph) [reply g3doc/DeveloperGuide.md]: Nice catch!

**Comment by joker-eph** [comment]
> Can we add links to the dialects mentioned in the examples?

I'd like to point to the website after I'm done with the auto-generation

**Pull request closed without merging**